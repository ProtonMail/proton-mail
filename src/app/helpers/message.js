/**
 * Get sender from message
 * @param {Object} message.Sender
 * @return {String} Name || Address
 */
export const getSender = ({ Sender = {} } = {}) => {
    const { Name = '', Address = '' } = Sender;
    return Name || Address;
};

/**
 * Get recipients list from message
 * @param {Array} message.TolList
 * @param {Array} message.BCCList
 * @param {Array} message.CCList
 * @return {Array} [Name || Address]
 */
export const getRecipients = ({ ToList = [], CCList = [], BCCList = [] } = {}) => {
    return [...ToList, ...CCList, ...BCCList].map(({ Address, Name }) => Name || Address);
};

/**
 * Get date from message
 * @param {Integer} message.Time
 * @return {Date}
 */
export const getDate = ({ Time = 0 } = {}) => new Date(Time * 1000);
