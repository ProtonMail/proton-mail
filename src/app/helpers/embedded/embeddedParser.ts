import { MessageExtended, Message } from '../../models/message';
import { escapeSrc, unescapeSrc, wrap } from '../dom';
import { Api } from '../../models/utils';
import { ENCRYPTED_STATUS } from '../../constants';
import { listInlineAttachments, getAttachment } from './embeddedFinder';
import { hasBlob, store, getBlob, BlobInfo } from './embeddedStoreBlobs';
import { get } from '../attachment/attachmentLoader';
import { Attachment } from '../../models/attachment';
import { noop } from 'proton-shared/lib/helpers/function';
import { wait } from 'proton-shared/lib/helpers/promise';
import { getMessageCIDs } from './embeddedStoreCids';
import { findEmbedded, srcToCID } from './embeddedUtils';
import { isInlineEmbedded, isEmbedded } from '../image';
import { AttachmentsDataCache } from '../../hooks/useAttachments';

const EMBEDDED_CLASSNAME = 'proton-embedded';

const wrapImage = (img: Element) => wrap(img, '<div class="image loading"></div>');

/**
 * Get the url for an embedded image
 */
export const getUrl = (node: Element) => {
    // If it's an inline embedded img, just return the src because that contains the img data.
    const src = node.getAttribute('data-embedded-img') || '';
    if (isInlineEmbedded(src)) {
        return src;
    }
    const cid = srcToCID(node);
    const { url = '' } = getBlob(cid);
    return url;
};

/**
 * Prepare embedded images in the document
 */
export const prepareImages = (message: MessageExtended, show: boolean, isReplyForward: boolean, isOutside: boolean) => {
    if (!message.document) {
        return;
    }

    let showEmbedded = message.showEmbeddedImages;

    const images = [...message.document.querySelectorAll('img[proton-src]')];

    images.forEach((image) => {
        const src = image.getAttribute('proton-src') || undefined;
        image.setAttribute('referrerPolicy', 'no-referrer');
        const attachment = getAttachment(message.data, src);

        console.log('prepareImages', image, src, attachment);

        if (!image.classList.contains(EMBEDDED_CLASSNAME)) {
            image.classList.add(EMBEDDED_CLASSNAME);
        }

        if (!image.parentElement) {
            return;
        }

        // check if the attachment exist before processing
        if (!(attachment && Object.keys(attachment).length > 0)) {
            /**
             * If the attachment does not exist and the proton-src attribute
             * starts with cid:, it's an embedded image that does not exist in the list of attachments,
             * or is not a valid image.
             * So remove the element from the DOM because it will not display anything useful anyway.
             */
            if (isEmbedded(src)) {
                image.parentElement.removeChild(image);
            }
            // If it's not an inline embedded image, it's either an embedded image or a remote image. So stop here.
            // Otherwise, continue so that the inline image is detected as an embedded image.
            if (!isInlineEmbedded(src)) {
                return;
            }
        }

        if (show) {
            image.setAttribute('data-embedded-img', src || '');
            /**
             * Since the image is supposed to be displayed, remove the proton-src attribute.
             * Then it will be parsed by the embeddedParser in the blob or cid direction.
             */
            image.removeAttribute('proton-src');

            // Auto load image inside a reply draft
            if (isReplyForward) {
                // `getUrl` may return undefined here because the embedded attachments have not yet been decrypted and put in the blob store.
                const url = getUrl(image);
                // only set it if it is defined, otherwise the unescapeSrc will add two src=""
                url && image.setAttribute('src', url);
                return;
            }

            // We don't need to add it outside
            if (!isOutside) {
                !image.parentElement.classList.contains('loading') && wrapImage(image);
                image.removeAttribute('src');
            }
            return;
        }

        showEmbedded = false;

        // Inline embedded images does not have an attachment.
        if (attachment) {
            image.setAttribute('alt', attachment.Name || '');
        }
    });

    return showEmbedded;
};

/**
 * launch and forget: we don't need to do anything with the result
 * wait a bit before disabling the invalidsignature modal
 * this allows the user to see the change icon popup.
 *
 * More elaborate explanation:
 * We're addressing a fairly rare UX thing here.
 * We want to avoid showing a popup saying the confirmmodal when the signature is invalid to often.
 * For instance, when embedding images you can see that the icon says the signature is invalid,
 * so we don't show this icon (as the user can know it before clicking).
 *
 * However, if you would click on the embedded attachment before it has downloaded the attachment, it will not show this icon.
 * So consider you clicking on this attachment when it didn't verify the attachment yet.
 * Then just after that the attachment loader downloaded the attachment and
 * verified it signature and sets invalidSignature.askAgain to false.
 * Then you don't know that this happened, but in this case you should get a popup.
 *
 * Note when thinking  this is just a race condition: also consider the case where you are clicking
 * on the icon and it shows the icon just before you click: it's not humanly possible to see that it
 * changed and is not valid. So even in that case we want to show the icon.
 */
const triggerSigVerification = (
    message: MessageExtended,
    attachments: Attachment[],
    api: Api,
    cache: Map<string, any>
) => {
    /*
     * launch and forget: we don't need to do anything with the result
     * wait a bit before disabling the invalidsignature modal
     * this allows the user to see the change icon popup.
     */
    Promise.all(
        attachments.map(async (attachment) => {
            await get(attachment, message, cache, api);
            await wait(1000);
            // invalidSignature.askAgain(message, attachment, false);
        })
    );
};

/**
 * It works on data-src attribute for this reason:
 * Don't set the src attribute since it's evaluated and cid:cid create an error (#3330)
 * NET::ERR_UNKNOWN_URL_SCHEME because src="cid:xxxx" is not valid HTML
 * This function expects the content to be properly unescaped later.
 */
const actionDirection: { [key: string]: (nodes: Element[], cid: string, url: string) => void } = {
    blob(nodes: Element[], cid: string, url: string) {
        nodes.forEach((node) => {
            console.log('blob', node, cid, url);
            // Always remove the `data-` src attribute set by the cid function, otherwise it can get displayed if the user does not auto load embedded images.
            node.removeAttribute('data-src');
            if (node.getAttribute('proton-src')) {
                return;
            }
            node.setAttribute('data-src', url);
            node.setAttribute('data-embedded-img', cid);
            node.classList.add(EMBEDDED_CLASSNAME);
        });
    },
    cid(nodes: Element[], cid: string) {
        nodes.forEach((node) => {
            node.removeAttribute('data-embedded-img');
            node.removeAttribute('src');
            node.setAttribute('data-src', `cid:${cid}`);
        });
    }
};

/**
 * Parse the content to inject the generated blob src
 */
export const mutateHTML = (message: Message = {}, document?: Element, direction: 'blob' | 'cid' = 'blob') => {
    if (!document) {
        return;
    }

    document.innerHTML = escapeSrc(document.innerHTML);

    console.log('mutateHTML before', getMessageCIDs(message), document.innerHTML);

    Object.keys(getMessageCIDs(message)).forEach((cid) => {
        const nodes = findEmbedded(cid, document);

        console.log('mutateHTML', cid, nodes);

        if (nodes.length) {
            const { url = '' } = getBlob(cid);

            (actionDirection[direction] || noop)(nodes, cid, url);
        }
    });

    console.log('mutateHTML after', document.innerHTML);

    document.innerHTML = unescapeSrc(document.innerHTML);

    console.log('mutateHTML after unescape', document.innerHTML);
};

export const decrypt = async (message: MessageExtended, api: Api, cache: AttachmentsDataCache) => {
    const list = listInlineAttachments(message);
    const attachments = list.map(({ attachment }) => attachment);
    // const show = message.showEmbeddedImages === true || mailSettings.ShowImages & SHOW_IMAGES.EMBEDDED;
    // const sigList = show ? list : list.filter(({ attachment }) => cache.has(attachment.ID));

    // For a draft if we close it before the end of the attachment upload, there are no keyPackets
    // pgp attachments do not have keypackets.

    const promises = list
        .filter(({ attachment }) => attachment.KeyPackets || attachment.Encrypted === ENCRYPTED_STATUS.PGP_MIME)
        .filter(({ cid }) => !hasBlob(cid))
        .map(async ({ cid, attachment }) => {
            const storeAttachement = store(message.data, cid);
            const buffer = await get(attachment, message, cache, api);
            return storeAttachement(buffer.data, attachment.MIMEType);
        });

    if (!promises.length) {
        // all cid was already stored, we can resolve
        triggerSigVerification(message, attachments, api, cache);
        return {};
    }

    await Promise.all(promises);

    // We need to trigger on the original list not after filtering: after filter they are just stored
    // somewhere else
    triggerSigVerification(message, attachments, api, cache);
    return list.reduce((acc, { cid }) => {
        acc[cid] = getBlob(cid);
        return acc;
    }, {} as { [cid: string]: BlobInfo });
};
