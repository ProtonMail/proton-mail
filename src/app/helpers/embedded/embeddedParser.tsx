import React from 'react';
import { render } from 'react-dom';
import { wait } from 'proton-shared/lib/helpers/promise';
import { Api } from 'proton-shared/lib/interfaces';
import { Attachment } from 'proton-shared/lib/interfaces/mail/Message';
import { isDraft } from 'proton-shared/lib/mail/messages';
import { Loader } from 'react-components';
import { MessageExtended, EmbeddedMap, MessageKeys, MessageVerification } from '../../models/message';
import { escapeSrc, unescapeSrc, wrap } from '../dom';
import { ENCRYPTED_STATUS } from '../../constants';
import { getAttachment, findEmbedded } from './embeddedFinder';
import { get } from '../attachment/attachmentLoader';
import { createBlob, readCID } from './embeddeds';
import { isInlineEmbedded, isEmbedded } from '../image';
import { AttachmentsCache } from '../../containers/AttachmentProvider';

export const EMBEDDED_CLASSNAME = 'proton-embedded';
export const EMBEDDED_CLASSNAME_LOADING = 'proton-embedded-loading';

const wrapImage = (img: Element) => {
    const element = wrap(img, `<div class="${EMBEDDED_CLASSNAME_LOADING} relative inline-block"></div>`);
    const loaderContainer = document.createElement('div');
    loaderContainer.classList.add('covered-absolute', 'flex');
    element.appendChild(loaderContainer);
    // Using a mini react to have the official loader
    render(<Loader className="center flex" />, loaderContainer);
};

/**
 * Prepare embedded images in the document
 */
export const prepareImages = (message: Partial<MessageExtended>, show: boolean) => {
    if (!message.document) {
        return;
    }

    const draft = isDraft(message.data);

    let showEmbedded = message.showEmbeddedImages;

    const images = [...message.document.querySelectorAll('img[proton-src]')];

    images.forEach((image) => {
        const src = image.getAttribute('proton-src') || undefined;
        image.setAttribute('referrerPolicy', 'no-referrer');
        const info = getAttachment(message.embeddeds, src);

        if (!image.classList.contains(EMBEDDED_CLASSNAME)) {
            image.classList.add(EMBEDDED_CLASSNAME);
        }

        if (!image.parentElement) {
            return;
        }

        // check if the attachment exist before processing
        if (!(info && Object.keys(info).length > 0)) {
            /**
             * If the attachment does not exist and the proton-src attribute
             * starts with cid:, it's an embedded image that does not exist in the list of attachments,
             * or is not a valid image.
             * So remove the element from the DOM because it will not display anything useful anyway.
             */
            if (isEmbedded(src)) {
                image.parentElement.removeChild(image);
            }
            // If it's not an inline embedded image, it's either an embedded image or a remote image. So stop here.
            // Otherwise, continue so that the inline image is detected as an embedded image.
            if (!isInlineEmbedded(src)) {
                return;
            }
        }

        if (show) {
            image.setAttribute('data-embedded-img', src || '');
            /**
             * Since the image is supposed to be displayed, remove the proton-src attribute.
             * Then it will be parsed by the embeddedParser in the blob or cid direction.
             */
            image.removeAttribute('proton-src');
            image.removeAttribute('src');

            if (!draft && !image.parentElement.classList.contains(EMBEDDED_CLASSNAME_LOADING)) {
                wrapImage(image);
            }

            return;
        }

        showEmbedded = false;

        // Inline embedded images does not have an attachment.
        if (info) {
            image.setAttribute('alt', info.attachment.Name || '');
        }
    });

    return showEmbedded;
};

/**
 * Remove an embedded attachment from the document
 */
export const removeEmbeddedHTML = (document: Element, attachment: Attachment) => {
    const cid = readCID(attachment);
    const nodes = document.querySelectorAll(
        `img[src="cid:${cid}"], img[data-embedded-img="cid:${cid}"], img[data-embedded-img="${cid}"]`
    );
    [...nodes].map((node) => node.remove());
};

/**
 * launch and forget: we don't need to do anything with the result
 * wait a bit before disabling the invalidsignature modal
 * this allows the user to see the change icon popup.
 *
 * More elaborate explanation:
 * We're addressing a fairly rare UX thing here.
 * We want to avoid showing a popup saying the confirmmodal when the signature is invalid to often.
 * For instance, when embedding images you can see that the icon says the signature is invalid,
 * so we don't show this icon (as the user can know it before clicking).
 *
 * However, if you would click on the embedded attachment before it has downloaded the attachment, it will not show this icon.
 * So consider you clicking on this attachment when it didn't verify the attachment yet.
 * Then just after that the attachment loader downloaded the attachment and
 * verified it signature and sets invalidSignature.askAgain to false.
 * Then you don't know that this happened, but in this case you should get a popup.
 *
 * Note when thinking  this is just a race condition: also consider the case where you are clicking
 * on the icon and it shows the icon just before you click: it's not humanly possible to see that it
 * changed and is not valid. So even in that case we want to show the icon.
 */
const triggerSigVerification = (
    verification: MessageVerification | undefined,
    messageKeys: MessageKeys,
    attachments: Attachment[],
    api: Api,
    cache: AttachmentsCache
) => {
    /*
     * launch and forget: we don't need to do anything with the result
     * wait a bit before disabling the invalidsignature modal
     * this allows the user to see the change icon popup.
     */
    void Promise.all(
        attachments.map(async (attachment) => {
            await get(attachment, verification, messageKeys, cache, api);
            await wait(1000);
            // invalidSignature.askAgain(message, attachment, false);
        })
    );
};

/**
 * It works on data-src attribute for this reason:
 * Don't set the src attribute since it's evaluated and cid:cid create an error (#3330)
 * NET::ERR_UNKNOWN_URL_SCHEME because src="cid:xxxx" is not valid HTML
 * This function expects the content to be properly unescaped later.
 */
const mutateHTML = (
    embeddeds: EmbeddedMap | undefined,
    document: Element | undefined,
    callback: (elements: Element[], cid: string) => void
) => {
    if (!embeddeds || !document) {
        return;
    }

    document.innerHTML = escapeSrc(document.innerHTML);
    embeddeds.forEach((_, cid) => callback(findEmbedded(cid, document), cid));
    document.innerHTML = unescapeSrc(document.innerHTML);
};

/**
 * Parse the content to inject the generated blob src
 */
export const mutateHTMLBlob = (embeddeds: EmbeddedMap | undefined, document: Element | undefined) => {
    mutateHTML(embeddeds, document, (elements, cid) => {
        const { url = '' } = embeddeds?.get(cid) || {};
        elements.forEach((element) => {
            // Always remove the `data-` src attribute set by the cid function, otherwise it can get displayed if the user does not auto load embedded images.
            element.removeAttribute('data-src');
            if (element.getAttribute('proton-src')) {
                return;
            }
            element.setAttribute('data-src', url);
            element.setAttribute('data-embedded-img', cid);
            element.classList.add(EMBEDDED_CLASSNAME);

            console.log('mutateHTML', element.parentElement?.parentElement);

            if (element.parentElement?.classList.contains(EMBEDDED_CLASSNAME_LOADING)) {
                console.log('replaceChild', element.parentElement.parentElement);
                element.parentElement.parentElement?.replaceChild(element, element.parentElement);
            }
        });
    });
};

/**
 * Parse the content to inject the cid
 */
export const mutateHTMLCid = (embeddeds: EmbeddedMap | undefined, document: Element | undefined) => {
    mutateHTML(embeddeds, document, (elements, cid) => {
        elements.forEach((element) => {
            element.removeAttribute('data-embedded-img');
            element.removeAttribute('src');
            element.setAttribute('data-src', `cid:${cid}`);
        });
    });
};

export const decrypt = async (
    message: MessageExtended,
    messageKeys: MessageKeys,
    api: Api,
    cache: AttachmentsCache
) => {
    // const show = message.showEmbeddedImages === true || mailSettings.ShowImages & SHOW_IMAGES.EMBEDDED;
    // const sigList = show ? list : list.filter(({ attachment }) => cache.has(attachment.ID));

    // For a draft if we close it before the end of the attachment upload, there are no keyPackets
    // pgp attachments do not have keypackets.

    const infos = [...(message.embeddeds?.values() || [])];

    const promises = infos
        .filter((info) => info.attachment.KeyPackets || info.attachment.Encrypted === ENCRYPTED_STATUS.PGP_MIME)
        .filter((info) => !info.url)
        .map(async (info) => {
            const buffer = await get(info.attachment, message.verification, messageKeys, cache, api);
            info.url = createBlob(info.attachment, buffer.data as Uint8Array);
        });

    const attachments = infos.map((info) => info.attachment);

    if (!promises.length) {
        // all cid was already stored, we can resolve
        triggerSigVerification(message.verification, messageKeys, attachments, api, cache);
        return;
    }

    await Promise.all(promises);

    // We need to trigger on the original list not after filtering: after filter they are just stored
    // somewhere else
    triggerSigVerification(message.verification, messageKeys, attachments, api, cache);
};
