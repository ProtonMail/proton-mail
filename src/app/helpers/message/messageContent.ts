import { MailSettings, Address } from 'proton-shared/lib/interfaces';
import { isPlainText } from 'proton-shared/lib/mail/messages';

import { Message } from 'proton-shared/lib/interfaces/mail/Message';
import { MessageExtended, PartialMessageExtended } from '../../models/message';
import { toText } from '../parserHtml';
import { findSender } from '../addresses';
import { textToHtml } from '../textToHtml';
import { parseInDiv } from '../dom';

export const getPlainTextContent = (message: PartialMessageExtended) => {
    return message.plainText || '';
};

export const getDocumentContent = (document: Element | undefined) => {
    const root = document?.querySelector('body') || document;
    return root?.innerHTML || '';
};

/**
 * Get current processed message document html content
 */
export const getContent = (message: PartialMessageExtended) => {
    if (isPlainText(message.data)) {
        return getPlainTextContent(message);
    }

    return getDocumentContent(message.document);
};

export const setPlainTextContent = (message: MessageExtended, content: string) => {
    message.plainText = content;
};

export const setDocumentContent = (document: Element | undefined, content: string) => {
    if (document) {
        document.innerHTML = content;
    } else {
        document = parseInDiv(content);
    }

    return document;
};

/**
 * Set current processed message document html
 */
export const setContent = (message: MessageExtended, content: string) => {
    if (isPlainText(message.data)) {
        setPlainTextContent(message, content);
    } else {
        message.document = setDocumentContent(message.document, content);
    }
};

export const exportPlainText = (content: string) => {
    /*
     * The replace removes any characters that are produced by the copying process (like zero width characters)
     * See: http://www.berklix.org/help/majordomo/#quoted we want to avoid sending unnecessary quoted printable encodings
     */
    return toText(content, true).replace(/\u200B/g, '');
};

/**
 * Generates/Gets the plaintext body from the message. If the message is not composed in plaintext, it will downconvert
 * the html body to plaintext if downconvert is set. If downconvert is disabled it will return false.
 */
export const getPlainText = (message: MessageExtended, downconvert: boolean) => {
    if (isPlainText(message.data)) {
        return getPlainTextContent(message);
    }

    if (!downconvert) {
        return undefined;
    }

    return exportPlainText(getContent(message));
};

/**
 * Convert the body of a message in plain text to an HTML version
 */
export const plainTextToHTML = (
    message: Message | undefined,
    plainTextContent: string | undefined,
    mailSettings: Partial<MailSettings> = {},
    addresses: Address[]
) => {
    const sender = findSender(addresses, message);
    return textToHtml(plainTextContent, sender?.Signature || '', mailSettings);
};

export const querySelectorAll = (message: Partial<MessageExtended> | undefined, selector: string) => [
    ...((message?.document?.querySelectorAll(selector) || []) as HTMLElement[]),
];
