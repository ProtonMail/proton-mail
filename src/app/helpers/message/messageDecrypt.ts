import { decryptMIMEMessage, decryptMessageLegacy } from 'pmcrypto';
import { MIME_TYPES } from 'proton-shared/lib/constants';
import { c } from 'ttag';

import { Message, MessageExtended } from '../../models/message';
import { convert } from '../attachment/attachmentConverter';
import { VERIFICATION_STATUS } from '../../constants';
import { getDate } from './messages';
import { AttachmentsDataCache } from '../../hooks/useAttachments';

const getVerifiedStatus = (pmcryptoVerified: number, publicKeys: any) => {
    const signedInvalid = VERIFICATION_STATUS.SIGNED_AND_INVALID;
    const signedPubkey = VERIFICATION_STATUS.SIGNED_NO_PUB_KEY;
    return !publicKeys.length && pmcryptoVerified === signedInvalid ? signedPubkey : pmcryptoVerified;
};

export const decryptMimeMessage = async (
    message: Message,
    privateKeys: any,
    publicKeys: any,
    attachmentsCache: AttachmentsDataCache
): Promise<MessageExtended> => {
    const headerFilename = c('Encrypted Headers').t`Encrypted Headers filename`;
    const sender = (message.Sender || {}).Address;

    const result = await decryptMIMEMessage({
        message: message.Body,
        messageDate: getDate(message),
        privateKeys,
        publicKeys,
        headerFilename,
        sender
    });

    const { body = c('Message empty').t`Message content if empty`, mimetype = MIME_TYPES.PLAINTEXT } =
        (await result.getBody()) || {};

    const verified = getVerifiedStatus(await result.verify(), publicKeys);

    const attachments = convert(message, await result.getAttachments(), verified, attachmentsCache);
    const encryptedSubject = await result.getEncryptedSubject();

    return { raw: body, data: { ...message, Attachments: attachments }, verified, encryptedSubject, mimetype };
};

export const decryptLegacyMessage = async (
    message: Message,
    privateKeys: any,
    publicKeys: any
): Promise<MessageExtended> => {
    const { data, verified: pmcryptoVerified = VERIFICATION_STATUS.NOT_SIGNED } = (await decryptMessageLegacy({
        message: message.Body,
        messageDate: getDate(message),
        privateKeys,
        publicKeys
    })) as any;

    const verified = getVerifiedStatus(pmcryptoVerified, publicKeys);

    return { raw: data, verified, publicKeys, privateKeys };
};
