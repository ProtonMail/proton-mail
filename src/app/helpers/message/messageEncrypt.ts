import { encryptMessage, PmcryptoKey } from 'pmcrypto';
import { MessageExtended } from '../../models/message';
import { getHTML } from './messages';

// Reference: Angular/src/app/message/factories/messageModel.js encryptBody

export const encryptBody = async (
    message: MessageExtended,
    privateKeys: PmcryptoKey[],
    publicKeys: PmcryptoKey[]
): Promise<MessageExtended> => {
    const { data } = await encryptMessage({
        data: getHTML(message) || '',
        publicKeys: [publicKeys[0]],
        privateKeys,
        format: 'utf8',
        compression: true
    });

    return { ...message, data: { ...message.data, Body: data } };
};
