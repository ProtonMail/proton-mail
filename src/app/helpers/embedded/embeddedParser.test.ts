import { prepareImages, mutateHTML } from './embeddedParser';
import { Message } from '../../models/message';
import { parseInDiv } from '../dom';

jest.mock('./embeddedFinder', () => ({
    getAttachment: jest.fn((message: Message, src: string) => {
        if (src === 'cid:embedded.jpg') {
            return { Name: 'embedded name' };
        }
    })
}));

jest.mock('./embeddedStoreCids', () => ({
    getMessageCIDs: jest.fn(() => ({ '-9e96b583@protonmail.com': {} }))
}));

jest.mock('./embeddedStoreBlobs', () => ({
    getBlob: jest.fn(() => ({ url: '123-234' }))
}));

describe('embeddedParser', () => {
    describe('prepareImages', () => {
        // Reference: Angular/test/specs/message/services/transformEmbedded.spec.js

        const EMBEDDED_CID = 'cid:embedded.jpg';
        const EMBEDDED_NAME = 'embedded name';
        const INLINE_EMBEDDED_URL = 'data:image/png;base64,iVBORw0KGgoA';
        const DOM = `
            <img proton-src="${EMBEDDED_CID}">
            <img proton-src="/remote.jpg">
            <img proton-src="${INLINE_EMBEDDED_URL}">
        `;

        const setup = (show = true) => {
            const message = { document: parseInDiv(DOM) };
            prepareImages(message, show, false, false);
            return message.document.querySelectorAll('img');
        };

        it('should add data-embedded-img for embedded img', () => {
            const imgs = setup();
            const embeddedImg = imgs[0];
            expect(embeddedImg.getAttribute('class')).toEqual('proton-embedded');
            expect(embeddedImg.getAttribute('data-embedded-img')).toEqual(EMBEDDED_CID);
            expect(embeddedImg.getAttribute('referrerpolicy')).toEqual('no-referrer');
        });

        it('should not add data-embedded-img for remote img', () => {
            const imgs = setup();
            const remoteImg = imgs[1];
            expect(remoteImg.getAttribute('data-embedded-img')).toBeNull();
        });

        it('should add src for inline embedded img', () => {
            const imgs = setup();
            const embeddedImg = imgs[2];
            expect(embeddedImg.getAttribute('class')).toEqual('proton-embedded');
            expect(embeddedImg.getAttribute('data-embedded-img')).toEqual(INLINE_EMBEDDED_URL);
            expect(embeddedImg.getAttribute('referrerpolicy')).toEqual('no-referrer');
        });

        it('should not add data-embedded-img for embedded img if show = false', () => {
            const imgs = setup(false);
            const embeddedImg = imgs[0];
            expect(embeddedImg.getAttribute('class')).toEqual('proton-embedded');
            expect(embeddedImg.getAttribute('data-embedded-img')).toBeNull();
            expect(embeddedImg.getAttribute('alt')).toEqual(EMBEDDED_NAME);
            expect(embeddedImg.getAttribute('referrerpolicy')).toEqual('no-referrer');
        });

        it('should not add data-embedded-img for remote img if show = false', () => {
            const imgs = setup(false);
            const remoteImg = imgs[1];
            expect(remoteImg.getAttribute('data-embedded-img')).toBeNull();
        });

        it('should not add data-embedded-img for inline embedded img if show = false', () => {
            const imgs = setup(false);
            const embeddedImg = imgs[2];
            expect(embeddedImg.getAttribute('class')).toEqual('proton-embedded');
            expect(embeddedImg.getAttribute('data-embedded-img')).toBeNull();
            expect(embeddedImg.getAttribute('alt')).toBeNull();
            expect(embeddedImg.getAttribute('referrerpolicy')).toEqual('no-referrer');
        });
    });

    describe('mutateHTML', () => {
        // Reference: Angular/test/specs/attachments/embeddedParser.spec.js

        const setup = (html: string, direction = 'blob') => {
            const message = { document: parseInDiv(html) };
            mutateHTML(message, direction);
            return message.document;
        };

        it('should parse embedded images and set src', async () => {
            const document = setup(
                '<img class="proton-embedded" data-src="cid:-9e96b583@protonmail.com" alt="embedded.jpg">'
            );
            expect(document.innerHTML).toMatchSnapshot();
        });

        it('should not parse embedded images and set src on proton-src', async () => {
            const document = setup(
                '<img class="proton-embedded" proton-src="cid:-9e96b583@protonmail.com" alt="embedded.jpg" >'
            );
            expect(document.innerHTML).toMatchSnapshot();
        });

        it('should always transform proton-src to src on direction cid', async () => {
            const document = setup(
                '<img class="proton-embedded" proton-src="cid:-9e96b583@protonmail.com" alt="embedded.jpg">',
                'cid'
            );
            expect(document.innerHTML).toMatchSnapshot();
        });

        it('should remove data-src when they are escaped with proton', async () => {
            const document = setup(
                '<img class="proton-embedded" proton-src="cid:-9e96b583@protonmail.com" data-src="cid:-9e96b583@protonmail.com" alt="embedded.jpg">'
            );
            expect(document.innerHTML).toMatchSnapshot();
        });
    });
});
