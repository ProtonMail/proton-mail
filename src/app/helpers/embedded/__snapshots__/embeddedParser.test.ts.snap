// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`embeddedParser mutateHTML should always transform proton-src to src on direction cid 1`] = `"<img class=\\"proton-embedded\\" proton-src=\\"cid:-9e96b583@protonmail.com\\" alt=\\"embedded.jpg\\" src=\\"cid:-9e96b583@protonmail.com\\">"`;

exports[`embeddedParser mutateHTML should not parse embedded images and set src on proton-src 1`] = `"<img class=\\"proton-embedded\\" proton-src=\\"cid:-9e96b583@protonmail.com\\" alt=\\"embedded.jpg\\">"`;

exports[`embeddedParser mutateHTML should parse embedded images and set src 1`] = `"<img class=\\"proton-embedded\\" alt=\\"embedded.jpg\\" src=\\"123-234\\" data-embedded-img=\\"-9e96b583@protonmail.com\\">"`;

exports[`embeddedParser mutateHTML should remove data-src when they are escaped with proton 1`] = `"<img class=\\"proton-embedded\\" proton-src=\\"cid:-9e96b583@protonmail.com\\" alt=\\"embedded.jpg\\">"`;
